### ISSUE TITLE
Adding `use Yii;` in the beginning of migration kills console

### ISSUE THREAD

**Issue by trejder** [yiisoft/yii2#8206]
To my extreme surprise, adding `use Yii;` in the beginning of migration file kills command-line (Git for Windows) very badly.

I think you may consider working-around this somehow.

I know, that `Yii` is available in migration and one does not / must not add `use Yii;` in the beginning of migration file. But, I'm pretty sure, you'd like to handle this situation a bit more gently, instead of firing user command-line console into one-way Mars mission, right? :>

**Comment by cebe** [comment]
what is the error you see? can you paste an example on how you added this? Do you have any php extensions installed like apc for example?

**Comment by trejder** [comment]
This is the migration code:

```php
<?php

use Yii;
use yii\db\Schema;
use yii\db\Migration;

class m150427_130904_test extends Migration
{
    public function up()
    {

    }

    public function down()
    {

    }
}
```

and this is the effect of running it (Polish Windows, but you get the picture, right?):

![bummer](https://cloud.githubusercontent.com/assets/2903392/7348211/e265fee6-ecef-11e4-858a-9ea2094b289d.png)

No APC or other cache, no magic, no nothing. Just an old, bit modified XAMPP running on Windows 7.

**Comment by samdark** [comment]
I'll try reproducing it.

**Comment by trejder** [comment]
Go ahead. Just added my list of extensions. Can provide you (zipped file, FTP link) with entire XAMPP package, I'm using, if you wish. Hope, this is something local, happening just for me, and not the general problem for all.

**Comment by mdmunir** [comment]
```
PHP Warning 'yii\base\ErrorException' with message 'The use statement with non-compound name 'Yii' has no effect'
```
IMO, its not yii bug, but php error. You cannot declare 2 class with same name.

**Comment by trejder** [comment]
Fine. But, I'm not talking, that this is Yii bug. This issue is about handling this gracefully. If we are required to manually add `use Yii;` to every class, where we want to use it, because it is not declared there automatically, then IMHO the same logic should be used in migrations.

In other words -- what is the reason for auto-declaring `Yii` class in migrations, if it is not auto-declared for other classes AFAIK.

**Comment by mdmunir** [comment]
> what is the reason for auto-declaring Yii class in migrations

Please, explain your question. i dont get it.

there is my generated migration. no `use Yii;` line.
```php
<?php

use yii\db\Schema;
use yii\db\Migration;

class m150427_131430_dont_do_anything extends Migration
{
    public function up()
    {

    }

    public function down()
    {
        echo "m150427_131430_dont_do_anything cannot be reverted.\n";

        return false;
    }
    
    /*
    // Use safeUp/safeDown to run migration code within a transaction
    public function safeUp()
    {
    }
    
    public function safeDown()
    {
    }
    */
}

```

**Comment by klimov-paul** [comment]
Migration classes are NOT under any namespace, thus using classes from global nemespace does not require 'use' statement.

**Comment by trejder** [comment]
@klimov-paul OK, fine... But, then again... No matter, what is causing this problem, I can't imagine any professional framework that causes command-line to blow if user do anything.

I reported this problem not to seek your explanation, what is causing this problem (but thank you, for explaining this to me), but to point you out, that you should at least consider, if there is any workaround for this? I know, that tracing or fixing this "problem" can't be beyond anything, you can do, but... then again, Yii2 is a professional framework and in a professional framework such extremely bad looking hangups shouldn't happen in any conditions.

**Comment by klimov-paul** [comment]
> causes command-line to blow if user do anything

PHP parse errors and warnings are out of the framework scope I am afraid.

> there is any workaround for this?

Are asking a wolkaround for PHP fatal error? I don't know any, sorry.

> such extremely bad looking hangups shouldn't happen in any conditions.

So if, for example, you miss a semicolon at the end of the statement or haven't enclose brackets correctly - framework should handle it somehow? Provide you a walk around? Are you serious?

Or there is something else I have missed?

**Comment by cebe** [comment]
@trejder yii has an errorhandler to deal with such errors, if your php crashes in some way this is a bug in php and yii can not do anything to fix this, this is another level.

**Comment by samdark** [comment]
I think the best thing to do in this case is to trace down error to minimal example and submit it as PHP bug. Of course, if you have time for it since it doesn't sound like easy debug session.

**Comment by trejder** [comment]
@klimov-paul Yes, I'm pretty serious. Please, don't make an equation between forgetting about missing semicolon and situation, when I should not declare `Yii` because it is already declared. Especially, when _in all other classes that I use in framework, I should declare it_! First is quite obvious -- anyone knows, that semicolon is required at the end of each line and nearly every good IDE can trace this down. Second is not so obvious at all and no IDE can trace it down. You're making two completely different situations looking like it would be similar.

@cebe I'm not sure, that Yii's error handler can trace redeclaration of class and deal with that in case of migration. If so, why then the console blows away on such obvious thing?

@samdark This is neither PHP nor Yii bug. This is just a very special situation, where migrations are run in global namespace and thus does not require declaring `Yii`, while all other classes and pieces of application are not run in global namespace and therefore requires such declaration.

All of you seems to be doing a really big problem out of something really small. In my opinion, a _workaround_, I'm asking for is to update `migrate` command's code, so it would insert simple comment to the beginning of each auto-generated migration:

> Keep in mind, that all migrations are run in global namespace and thus you shouldn't declare Yii here. It is already declared in global namespace and redeclaring it may give an unpredictable results.

This would "solve" the problem once and for all. Even those programmer, like me, for which -- against to what @klimov-paul says -- this situation isn't obvious, would be informed about possible danger of redeclaring `Yii` class inside migration.

I have a very bad feeling, that you're, again and again, keep forgetting that Yii2 users are not always so experienced as you are and not always posses such knowledge as you are. For you, the fact that migrations are run in global namespace, is pretty obvious. For me it wasn't. And due to lack of above proposed comment and due to lack of my knowledge I wasted over half an hour on tracing what is going on and why my console gets fired hell?

**Comment by mdmunir** [comment]
Just switch off debug mode.
> would be informed about possible danger of redeclaring Yii class inside migration.

Not only in migration file, also in view file and all file that not using namespace. Not only `Yii` class. but also all global class e.g. `PDO`, `ReflectionClass`, `Exception`, etc, etc.

**Comment by samdark** [comment]
@trejder PHP crashes so it's PHP bug. Normally it should have thrown an fatal error which is handled well by Yii.

**Comment by samdark** [comment]
Just checked it. It's not even PHP bug, it's your environment. That's what I have on Windows:

![Console error](https://cloud.githubusercontent.com/assets/47294/7366741/02f192f2-eda3-11e4-8bfd-2714a1252cbc.png)

**Comment by trejder** [comment]
Everything you have shown so far seems proving, that this is my local problem only. However, I found it hard to believe. Mainly because:

- I haven't changed _anything_ in my XAMPP structure, `php.ini` content etc. for past many, many months, I'm actually still using old-good 1.8.1 version of XAMPP (versioned  without compliance to PHP version),

- all metioned problems started about month ago, when I started coding in Yii2 and happens _only_ in Yii2-powered apps -- I can't reproduce any of above mentioned problems in any of my Yii1-powered projects.

I have asked [a Stack Overflow question](http://stackoverflow.com/q/30039929/1469208) on this matter -- in case someone would like to follow this.
